fix llvm build erros

peppergrayxyz opened the pull request:
fixes #146, #147 and #148

peppergrayxyz committed:
check if compiler has command -print-multi-os-directory

Closes: #146
Signed-off-by: Pepper Gray <hello@peppergray.xyz>

superm1 commented on efi/meson.build:
```diff
@@ -37,6 +37,9 @@ endif
 efi_libdir = get_option('efi-libdir')
 if efi_libdir == ''
+  if not meson.get_compiler('c').has_argument('-print-multi-os-directory')
+    error('efi-libdir is not set')
```
This error doesn't make sense for the test.

peppergrayxyz replied:
Should we put something like "compiler doesn't support -print-multi-os-directory"? I'm also not sure if this should be an error or a warning, and then default to '/usr/lib'. What do you think is a good exception handling strategy here?

superm1 replied:
Hmm. I think warning and fall back is ideal considering there is already some fallback paths right now.

superm1 reviewed (commented):
Changes look good to me; but one more thing that I think would be worthwhile is to add a CI target for compiling with llvm so that we don't regress anything like this again in the future.

Maybe even using an Alpine container :P

peppergrayxyz committed:
probe if objcopy supports efi-app

instead of hard coding (gnu)objcopy version numbers,
create an efi probe and check if objcopy is able to
create an efi-app elf file from it.

Signed-off-by: Pepper Gray <hello@peppergray.xyz>

peppergrayxyz committed:
refactor local_crt and local_lds handling

Signed-off-by: Pepper Gray <hello@peppergray.xyz>

peppergrayxyz committed:
check system aligned lds and crt0

Signed-off-by: Pepper Gray <hello@peppergray.xyz>

peppergrayxyz committed:
rework ci

- add cross archicture alpine build
- add legacy binuitls ubuntu build

Signed-off-by: Pepper Gray <hello@peppergray.xyz>

peppergrayxyz committed:
probe if crt0 embedds a PE header

fall back to local crt0 if objcopy cannot create pefiles
and system crt0 does not embedd a PE header

Signed-off-by: Pepper Gray <hello@peppergray.xyz>

peppergrayxyz committed:
add PE header

Signed-off-by: Pepper Gray <hello@peppergray.xyz>

peppergrayxyz committed:
fix crt0-efi-loongarch64

clang build fails with:
```
../efi/crt0/crt0-efi-loongarch64.S:199:7: error: expected newline
 .end _start
```
align with upstream

Signed-off-by: Pepper Gray <hello@peppergray.xyz>

peppergrayxyz committed:
remove sse mfpmath from x86 build

building on alpine llvm fails with error:
```
error: the 'sse' unit is not supported with this instruction set
```
due to `mfpmath` being implicitly set to `see`

set `mfpmath` to `387`

Signed-off-by: Pepper Gray <hello@peppergray.xyz>

peppergrayxyz committed:
use --rtlib for -print-libgcc-file-name

Signed-off-by: Pepper Gray <hello@peppergray.xyz>
